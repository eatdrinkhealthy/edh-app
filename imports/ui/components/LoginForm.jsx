// @flow
import React from "react";
import { Field, reduxForm } from "redux-form";
import classNames from "classnames";

const requiredUsername = value => (value ? undefined : `Username / Email is a required field.`);
const requiredPassword = value => (value ? undefined : `Password is a required field.`);

const renderInputError = (inputId: string, error: string) => (
  <div id={`${inputId}Error`} className="dark-red">
    {error}
  </div>
);

type IRenderInputProps = {
  inputId: string,
  autoFocus: boolean,
  label: string,
  type: string,
  input: {},          // redux-form flow type
  meta: {             // redux-form flow type
    touched: boolean, // eslint-disable-line react/no-unused-prop-types
    error: string,    // eslint-disable-line react/no-unused-prop-types
  },
};

const renderInput = (
  { input, inputId, autoFocus, label, type, meta: { touched, error } }: IRenderInputProps,
) => (
  /* eslint-disable jsx-a11y/no-autofocus */
  <div>
    <label
      className={classNames({ "dark-red": touched && !!error })}
      htmlFor={inputId}
    >
      {label}
    </label>
    <input
      {...input}
      id={inputId}
      className={classNames({ "dark-red": touched && !!error })}
      type={type}
      autoFocus={autoFocus}
    />
    {touched && (error && renderInputError(inputId, error))}
  </div>
  /* eslint-enable jsx-a11y/no-autofocus */
);

export type ILoginFormValues = {
  usernameEmail: string,
  loginPassword: string,
};

type ILoginFormProps = {
  handleSubmit: (ILoginFormValues) => void,
  invalid: boolean,
};

// NOTE: when we call the LoginForm component, we pass our submit handler as an
// onSubmit prop, and redux-form passes a submit function in as handleSubmit
const LoginFormComponent = ({ handleSubmit, invalid }: ILoginFormProps) => (
  <form onSubmit={handleSubmit}>
    <Field
      inputId="usernameEmail"
      name="usernameEmail"
      type="text"
      label="Username / Email"
      component={renderInput}
      validate={requiredUsername}
      autoFocus
    />
    <Field
      inputId="loginPassword"
      name="loginPassword"
      type="password"
      label="Password"
      component={renderInput}
      validate={requiredPassword}
    />
    <input
      type="submit"
      value="Login"
      disabled={invalid}
    />
  </form>
);

const onSubmitSuccess = (result, dispatch, props) => {
  // on successful submit, clear the password fields
  props.change("loginPassword", "");
  props.untouch("loginPassword");     // prevents showing validation error

  // set focus on usernameEmail field
  // NOTE: using document.querySelector is not a React standard (works here, not in jest)
  const usernameEmailElem = document.querySelector("input#usernameEmail");
  if (usernameEmailElem) {
    usernameEmailElem.focus();
  }
};

const LoginForm = reduxForm({
  form: "LoginForm",
  onSubmitSuccess,
})(LoginFormComponent);

export default LoginForm;
