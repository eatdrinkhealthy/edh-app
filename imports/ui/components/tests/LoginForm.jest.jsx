// @flow
/* eslint-env jest */
/* eslint-disable func-names, prefer-arrow-callback, no-unused-expressions */
/* eslint-disable import/no-extraneous-dependencies */
import React from "react";
import { createStore } from "redux";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import LoginForm from "../LoginForm";
import mountFormWithInputs from "../../../utils/tests/mountFormWithInputs";
import appReducer from "../../../state/reducers";

describe("<LoginForm />", function () {
  const testStore = createStore(appReducer);

  // NOTE: we actually pass handleSubmit to the component, but redux-form calls onSubmit
  const props = {
    onSubmit: jest.fn(),
  };

  it("matches render snapshot, with form validation errors", function () {
    // render form with no input values to display validation errors
    const tree = renderer.create(
      <Provider store={testStore}>
        <LoginForm onSubmit={() => {}} />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it.skip("should set username / email required validation error, when not provided", function () {

  });

  it.skip("should set password required validation error, when not provided", function () {

  });

  it("should call onSubmit on submit with form field values", function () {
    const wrapper = mountFormWithInputs(
      <LoginForm {...props} />,
      { usernameEmail: "user12", loginPassword: "user12pw" },
      testStore,
    );

    wrapper.find("input[type='submit']").simulate("submit");
    expect(props.onSubmit).toHaveBeenCalledWith(
      { usernameEmail: "user12", loginPassword: "user12pw" },
      expect.anything(),
      expect.anything(),
    );
  });

  it("should clear password on successful submit", function () {
    const wrapper = mountFormWithInputs(
      <LoginForm {...props} />,
      { usernameEmail: "user12", loginPassword: "user12pw" },
      testStore,
    );
    const usernameEmailNode = wrapper.find("input#usernameEmail");
    const passwordNode = wrapper.find("input#loginPassword");

    // give focus to password input (like a user would do before submit)
    // $FlowFixMe
    passwordNode.get(0).focus();
    wrapper.find("input[type='submit']").simulate("submit");

    expect(usernameEmailNode.props().value).toEqual("user12");
    expect(passwordNode.props().value).toEqual("");
    // NOTE: document.querySelector called in onSubmitSuccess works in browser but not in jest
    // expect(usernameEmailNode.get(0)).toEqual(document.activeElement);
  });

  it("should set focus to username input on render", function () {
    const wrapper = mountFormWithInputs(
      <LoginForm {...props} />,
      { usernameEmail: "user12", loginPassword: "user12pw" },
      testStore,
    );

    expect(wrapper.find("input#usernameEmail").get(0)).toBe(document.activeElement);
  });
});
