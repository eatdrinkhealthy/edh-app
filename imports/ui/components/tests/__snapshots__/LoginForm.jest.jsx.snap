// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> matches render snapshot, with form validation errors 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <label
      className=""
      htmlFor="usernameEmail"
    >
      Username / Email
    </label>
    <input
      autoFocus={true}
      className=""
      id="usernameEmail"
      name="usernameEmail"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      type="text"
      value=""
    />
  </div>
  <div>
    <label
      className=""
      htmlFor="loginPassword"
    >
      Password
    </label>
    <input
      autoFocus={undefined}
      className=""
      id="loginPassword"
      name="loginPassword"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      type="password"
      value=""
    />
  </div>
  <input
    disabled={true}
    type="submit"
    value="Login"
  />
</form>
`;
